### PR TITLE
fix(mme): Adding write of MME UE state for NAS uplink data ind handlers

### LINKED ISSUE #8980: [agw] MME non-sanity tests with service restart are broken
### Your Environment

- **Version:** master
- **Affected Component:** Access Gateway
- **Affected Subcomponent:** MME / S1AP tests
- **Deployment Environment:** Vagrant (AGW)

### Describe the Issue

- After introducing filtering of MME UE context writes for only registered users on #6945, some of the non-sanity tests with MME restart do not pass, this is because there will not be any written context to redis for MME task only till UE completes attach procedure.
- E.g. ICS drop with MME restart test becomes stuck, as after service restart MME cannot complete implicit detach due to not existent MME UE context.

**To Reproduce**
- Run any non-sanity test with MME restart (e.g. `test_attach_ics_drop_with_mme_restart.py`)
- Wait for S1AP test to get stuck, also S1AP UE and SPGW UE context will be stale on redis db.

**Expected behavior**
- Non-sanity tests with MME restart should be successful.

### LINKED ISSUE #9004: [agw] Non-sanity tests are unsuccessful due to ZMQ loop assert
### Your Environment

- **Version:** master
- **Affected Component:** Access Gateway
-  **Affected Subcomponent:** S1AP tester, MME
- **Deployment Environment:** Vagrant (AGW)

### Describe the Issue

- Some of the tests part of the `nonsanity` step cause a MME restart, due to assertion ZMQ thread context assert:
```
Sep 08 19:10:34 magma-dev-focal mme[339165]: mme: src/zlistx.c:172: zlistx_add_end: Assertion `item' failed.
Sep 08 19:10:37 magma-dev-focal systemd[1]: magma@mme.service: Main process exited, code=killed, status=6/ABRT
Sep 08 19:10:37 magma-dev-focal systemd[1]: magma@mme.service: Failed with result 'signal'.
Sep 08 19:10:42 magma-dev-focal systemd[1]: magma@mme.service: Scheduled restart job, restart counter is at 1.
Sep 08 19:10:42 magma-dev-focal systemd[1]: Stopped Magma OAI MME service.
Sep 08 19:10:45 magma-dev-focal systemd[1]: Starting Magma OAI MME service...
```

List of tests:
- `s1aptests/test_no_esm_information_rsp_with_mme_restart`
- `s1aptests/test_no_smc_with_mme_restart_reattach.py`
- `s1aptests/test_no_attach_complete_with_mme_restart.py`

**Expected behavior**
- Given s1aptests should pass without zmq assertion on thread loop

### LINKED ISSUE #9605: Add MME_APP/NAS state conversion unit tests
MME APP state converter (https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/tasks/mme_app/mme_app_state_converter.cpp) needs code coverage by adding unit testing. These are well-contained functions that don't have any dependency to mock/include.

For reference and an example of a state converter unit test, please refer to: https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/test/spgw_task/test_spgw_state_converter.cpp

### PR DESCRIPTION
**Pull request by ardzoht**
## Summary

- Closes #8980 and #9004
- MME UE context writes only on registered mode are not sufficient, as it needs coverage for uncompleted attach procedure handlers (e.g ICS not completed, Authentication without response)
- This extends MME UE writes for NAS uplink procedure exchanges, while maintaining UE context writes only for registered subs filter.

## Test Plan

- Running `make integ_test` as sanity
- All nonsanity restart tests are passing
MME UE writes without Registered filter (running active idle test): 13
MME UE writes with Registered filter and NAS procedures (running active idle test): 9

- Added unit tests for EMM context state conversion and timers invalid ID tests.

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by ulaskozat** [comment]
can you run all non-sanity restart tests?

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_state.cpp [hunk 1]
```diff
@@ -60,13 +60,15 @@ hash_table_ts_t* get_mme_ue_state() {
...
-void put_mme_ue_state(mme_app_desc_t* mme_app_desc_p, imsi64_t imsi64) {
+void put_mme_ue_state(
+    mme_app_desc_t* mme_app_desc_p, imsi64_t imsi64, bool force_ue_write) {
   if (MmeNasStateManager::getInstance().is_persist_state_enabled()) {
     if (imsi64 != INVALID_IMSI64) {
       ue_mm_context_t* ue_context = nullptr;
       ue_context =
           mme_ue_context_exists_imsi(&mme_app_desc_p->mme_ue_contexts, imsi64);
-      if (ue_context && ue_context->mm_state == UE_REGISTERED) {
+      if ((ue_context && !force_ue_write ||
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_state.cpp]: missing parenthesis before `||`?

File: lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.c [hunk 2]
```diff
@@ -226,8 +226,8 @@ status_code_e sgw_cm_remove_bearer_context_information(
             "Failed to free imsi64 from imsi_ue_context_htbl \n");
         return temp;
       }
-      delete_spgw_ue_state(imsi64);
     }
+    delete_spgw_ue_state(imsi64);
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.c]: this does not look correct to me. Should not we delete UE state when where is no more bearer contexts?

**Review by ulaskozat — CHANGES_REQUESTED** [review]
see review comments.

**Comment by ardzoht** [comment]
> can you run all non-sanity restart tests?

@ulaskozat I ran all nonsanity suite, however, there are a couple of tests that are failing due to an assertion on the zmq thread (probably related to timer?) Also, these are unrelated with the write-only for registered UEs, as this assert still happens without the filter check on the Registered EMM state.

This fixes `test_attach_ics_drop_with_mme_restart` and `test_attach_ics_failure_with_mme_restart`, for the other tests that are failing, I created #9004 to track, I will take a look into these.

File: lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.c (hunk 2 above)
Comment (ardzoht) [reply lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.c]: You're correct, I was trying to resolve SPGW UE staleness but after checking that is due to the other errors on zmq assertion.

**Comment by ardzoht** [comment]
> > can you run all non-sanity restart tests?
> 
> @ulaskozat I ran all nonsanity suite, however, there are a couple of tests that are failing due to an assertion on the zmq thread (probably related to timer?) Also, these are unrelated with the Registered write-only, as I tested a binary without the filter and the assertion still happens.
> 
> This fixes `test_attach_ics_drop_with_mme_restart` and `test_attach_ics_failure_with_mme_restart`, for the other tests that are failing, I created #9004 to track, I will take a look into these.

File: lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.c (hunk 2 above)
Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/sgw/sgw_context_manager.c]: can you update the summary?

**Comment by ulaskozat** [comment]
What about timer expirations? We no longer receive messages for those.

File: lte/gateway/c/core/oai/tasks/mme_app/mme_app_state.cpp [hunk 3]
```diff
@@ -60,13 +60,15 @@ hash_table_ts_t* get_mme_ue_state() {
...
-void put_mme_ue_state(mme_app_desc_t* mme_app_desc_p, imsi64_t imsi64) {
+void put_mme_ue_state(
+    mme_app_desc_t* mme_app_desc_p, imsi64_t imsi64, bool force_ue_write) {
   if (MmeNasStateManager::getInstance().is_persist_state_enabled()) {
     if (imsi64 != INVALID_IMSI64) {
       ue_mm_context_t* ue_context = nullptr;
       ue_context =
           mme_ue_context_exists_imsi(&mme_app_desc_p->mme_ue_contexts, imsi64);
-      if (ue_context && ue_context->mm_state == UE_REGISTERED) {
+      if ((ue_context && !force_ue_write) ||
```
Comment (lionelgo) [inline lte/gateway/c/core/oai/tasks/mme_app/mme_app_state.cpp]: isn't it 
if (ue_context && force_ue_write) ?

Comment (ulaskozat) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_app_state.cpp]: +1

Comment (ardzoht) [reply lte/gateway/c/core/oai/tasks/mme_app/mme_app_state.cpp]: good catch, I was using this as reverse (filter_ue_write) but renamed it and forgot to change the condition, will update

File: lte/gateway/c/core/oai/tasks/nas/esm/esm_data_context.c [hunk 4]
```diff
@@ -47,15 +47,6 @@ void free_esm_bearer_context(esm_ebr_context_t* esm_ebr_context) {
...
       nas_timer_stop(&(esm_ebr_context->timer));
-      /*
-       * Release the retransmisison timer parameters
-       */
-      if (esm_ebr_timer_data) {
-        if (esm_ebr_timer_data->msg) {
-          bdestroy_wrapper(&esm_ebr_timer_data->msg);
-        }
-        free_wrapper((void**) &esm_ebr_timer_data);
-      }
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/tasks/nas/esm/esm_data_context.c]: why is this not needed anymore? Line 48 should be `esm_ebr_timer_data_t* esm_ebr_timer_data =esm_ebr_context->args` and we should keep the lines from 50 through 58, I think.

File: lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+  nas_emm_auth_proc_t state_nas_emm_auth_proc;
+
+  NasStateConverter::proto_to_nas_emm_auth_proc(
+      auth_proc_proto, &state_nas_emm_auth_proc);
+
+  EXPECT_EQ(state_nas_emm_auth_proc.T3460.id, NAS_TIMER_INACTIVE_ID);
+}
+
+TEST_F(NasStateConverterTest, TestNasEmmSmcProcSetInactiveT3460) {
+  oai::SmcProc smc_proc_proto;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp:166]: should not you fill in smc_proc_proto and define an expected state_nas_emm_smc_proc? Then do memory comparison?

File: lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+  attach_proc_proto.clear_esm_msg_out();
+  magma::lte::NasStateConverter::proto_to_nas_emm_attach_proc(
+      attach_proc_proto, &state_nas_emm_attach_proc);
+
+  EXPECT_EQ(state_nas_emm_attach_proc.T3450.id, NAS_TIMER_INACTIVE_ID);
+}
+
+TEST_F(NasStateConverterTest, TestNasEmmAuthProcSetInactiveT3460) {
+  oai::AuthProc auth_proc_proto;
+  nas_emm_auth_proc_t state_nas_emm_auth_proc;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp:157]: define values of `auth_proc_proto` and define an `expected_state_nas_emm_auth_proc`. Make a memory comparison with EXPECT_EQ.

File: lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+  esm_context_proto.clear_esm_ebr_timer_data();
+  NasStateConverter::proto_to_esm_ebr_context(
+      esm_context_proto, &state_esm_ebr_context);
+
+  EXPECT_EQ(state_esm_ebr_context.timer.id, NAS_TIMER_INACTIVE_ID);
+}
+
+TEST_F(NasStateConverterTest, TestNasEmmAttachProcSetInactiveT3450) {
+  oai::AttachProc attach_proc_proto;
+  nas_emm_attach_proc_t state_nas_emm_attach_proc;
```
Comment (ulaskozat) [inline lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp:145]: please see my comments about filling up the protos, defining expected data structure and doing a memcmp.

File: lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp (hunk 5 above)
Comment (ardzoht) [reply lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp:166]: I was thinking for these to test only the expectation of these functions to set the according timers to inactive, and for the actual `TestEmmContextConversion` to test these, I think we can add separate tests for the memory comparisons.

File: lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp (hunk 6 above)
Comment (ardzoht) [reply lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp:157]: Same as my previous comment, I wanted these tests to be self contained on what they actually test, I will add separate tests for the memory comparison tests.

Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp:157]: it maybe an overkill if timer id will be already tested by the other test. we should avoid adding tests that are strictly subset of another test.

Comment (ulaskozat) [reply lte/gateway/c/core/oai/test/nas/test_nas_converter.cpp:157]: based on our offline discussion, let's keep them as you specified.

File: lte/gateway/c/core/oai/tasks/nas/esm/esm_data_context.c (hunk 4 above)
Comment (ardzoht) [reply lte/gateway/c/core/oai/tasks/nas/esm/esm_data_context.c]: Updated this

**Review by ulaskozat — APPROVED** [review]
LGTM

**Pull request merged**